pkg/kubelet/pleg/generic.go

```diff
@@ -218,36 +226,59 @@ func (g *GenericPLEG) relist() {
 		needsReinspection = make(map[types.UID]*kubecontainer.Pod)
 	}
 
+	// update PodCache and PodRecords
+	g.updatePodCacheByWorkers(eventsByPodID, needsReinspection)
+
+	if g.cacheEnabled() {
+		// reinspect any pods that failed inspection during the previous relist
+		if len(g.podsToReinspect) > 0 {
+			glog.V(5).Infof("GenericPLEG: Reinspecting pods that previously failed inspection")
+			for pid, pod := range g.podsToReinspect {
+				if err := g.updateCache(pod, pid); err != nil {
+					glog.Errorf("PLEG: pod %s/%s failed reinspection: %v", pod.Name, pod.Namespace, err)
+					needsReinspection[pid] = pod
+				}
+			}
+		}
+
+		// Update the cache timestamp.  This needs to happen *after*
+		// all pods have been properly updated in the cache.
+		g.cache.UpdateTime(timestamp)
+	}
+
+	// make sure we retain the list of pods that need reinspecting the next time relist is called
+	g.podsToReinspect = needsReinspection
+}
+
+func (g *GenericPLEG) updatePodCacheByWorkers(eventsByPodID map[types.UID][]*PodLifecycleEvent,
+	needsReinspection map[types.UID]*kubecontainer.Pod) {
 	// If there are events associated with a pod, we should update the
 	// podCache.
-	for pid, events := range eventsByPodID {
+	doUpdateCache := func(pid types.UID) {
+		events := eventsByPodID[pid]
 		pod := g.podRecords.getCurrent(pid)
 		if g.cacheEnabled() {
 			// updateCache() will inspect the pod and update the cache. If an
 			// error occurs during the inspection, we want PLEG to retry again
 			// in the next relist. To achieve this, we do not update the
 			// associated podRecord of the pod, so that the change will be
 			// detect again in the next relist.
-			// TODO: If many pods changed during the same relist period,
-			// inspecting the pod and getting the PodStatus to update the cache
-			// serially may take a while. We should be aware of this and
-			// parallelize if needed.
 			if err := g.updateCache(pod, pid); err != nil {
 				glog.Errorf("PLEG: Ignoring events for pod %s/%s: %v", pod.Name, pod.Namespace, err)
-
 				// make sure we try to reinspect the pod during the next relisting
 				needsReinspection[pid] = pod
-
-				continue
+				return
 			} else if _, found := g.podsToReinspect[pid]; found {
 				// this pod was in the list to reinspect and we did so because it had events, so remove it
 				// from the list (we don't want the reinspection code below to inspect it a second time in
 				// this relist execution)
 				delete(g.podsToReinspect, pid)
```

Comment:
Is it safe to delete item in different goroutines? Even though we know that the routines won't delete the same pod, but I'm not sure whether that is thread safe.